### ISSUE TITLE
Undo/Redo

### ISSUE THREAD

**Issue by arvind** [uwdata/lyra#236]
One way to support undo/redo is to make Lyra 2's model Immutable. Changes to the model can then be tracked as a state in a [history stack](http://www.macwright.org/2015/05/18/practical-undo.html). We will want a more granular API (e.g., with debouncing/throttling) such that we don't flood the state stack with small changes to properties for example.

**Comment by kadamwhite** [comment]
Potentially-useful resource: http://www.macwright.org/2015/05/18/practical-undo.html

**Comment by kadamwhite** [comment]
How much history do we want to maintain in the undo/redo queue, @arvind ?

**Comment by kadamwhite** [comment]
Also we'll probably want to support global shortcuts, https://github.com/ccampbell/mousetrap looks like it has a nicer API than a DIY solution but is now a bit outdated

**Comment by deathbearbrown** [comment]
**UPDATE ON UNDO REDO**
Redux Recipe for implementing undo redo:
http://redux.js.org/docs/recipes/ImplementingUndoHistory.html 

2 ways to structure history state: 
https://github.com/omnidan/redux-undo#initial-state-and-history

```
// this
{
    past: [{}, {}],
    present: {
        marks: [],
        vega: {},
        etc: 'etera'
    },
    future: []
}
// or this (standard, easy????)
{
    marks: {
        past: [[], []],
        present: [[]],
        future: []
    },
    vega: {},
    etc: 'etera'
}
```

The ideal situation is that when you hit undo, it rolls back redux to the previous state, then re-parses the view for vega. 

In both potential state structures shown above, there is an issue with undoing changes that are “less important” or intermediate like moving things around.

For example, moving a mark around the scene will update redux for every pixel moved and thus create a history state every time the value is updated. 

Top version is easier to coordinate changes between sibling stores, e.g. marks and signals, which depend on coordination; harder to omit things from the undoable state.

Using the undo-redo addon that is in the current PR #307  
https://github.com/omnidan/redux-undo
The reducers are wrapped in redux to add the state related structure (past, present, future). There is an option to set things to not be tracked. 

You could potentially pull out all "non trackables" into their own map in redux, and then create some sort of logic to tie everything in together.  Or you can modify the undo-redo addon to look for "untrackables" in a more granular way.

Whatever the case, a lot of thought needs to go into this before implementation. 

One POTENTIAL thing to look at is smart actions, though this is not something that was created to solve this issue. https://github.com/stephan83/redux-smart-action  It might be a tool used to get to a viable solution.  Or it might be useless.  Why is there no shrug emoji?

**Comment by arvind** [comment]
Implemented with #387!
